### PR TITLE
Rework CI configuration and bump versions

### LINKED ISSUE #31: Improvements for continuous integration
* Update to the latest version of go
* Test `getroughtime` against instances of `testserver` in a container
* Add some go linter

### PR DESCRIPTION
**Pull request by MDr164**
This bumps the versions of the used actions and also adds more steps. It now runs race condition and coverage tests as well and also runs a linter over the codebase. Resolves #31

### CODE REVIEW

**Comment by MDr164** [comment]
This PR still lacks a linter configuration, I'll add one with reasonable defaults in due time.

**Comment by cjpatton** [comment]
> This PR still lacks a linter configuration, I'll add one with reasonable defaults in due time.

`Build and Test / Golangci Lint` found some issues. Would you mind fixing them here?

**Comment by MDr164** [comment]
> > This PR still lacks a linter configuration, I'll add one with reasonable defaults in due time.
> 
> `Build and Test / Golangci Lint` found some issues. Would you mind fixing them here?
> 
> 

Yes sure, I'll do that :+1:

File: cmd/getroughtime/main.go
Comment (cjpatton) [inline cmd/getroughtime/main.go:43]: Did the linter suggest this? This change seems fine but more opinionated than I would expect a linter to be.

File: cmd/testserver/main.go
Comment (cjpatton) [inline cmd/testserver/main.go:39]: Same question as above: Did the linter suggest making these non-global?

**Review by cjpatton — COMMENTED** [review]
@mdr164 this looks ready for review.

File: cmd/getroughtime/main.go
Comment (MDr164) [reply cmd/getroughtime/main.go:43]: Here is a list of all golangci-lint available linter, one of those is the gochecknoglobals linter which got active locally. I can change those back. Maybe we should decide on a configuration to use (or just keep the defaults). Does Cloudflare use a common linter configuration across Go projects?

Comment (cjpatton) [reply cmd/getroughtime/main.go:43]: Thanks for the explanation, I'm happy to take this suggestion.

We don't have a standard linter config, at least for open source projects.

File: .github/workflows/build_and_test.yml
Comment (cjpatton) [inline .github/workflows/build_and_test.yml:1]: @armfazh can you take a look at this workflow? Anything you'd port over from CIRCL or does it look good as is?

File: cmd/getroughtime/main.go
Comment (cjpatton) [inline cmd/getroughtime/main.go:74]: Here and below:
```suggestion
			logger.Fatalf("Public key decode error: %s\n", err)
```

File: cmd/getroughtime/main.go
Comment (MDr164) [reply cmd/getroughtime/main.go:43]: It is indeed an opinionated linter, hence it is not part of the default set of linters. Additionally one of the commands in this repo already followed this schema, so figured I might as apply it to the others as well. Let me know if you got any preference here.

File: recipes/alerter.go
Comment (cjpatton) [inline recipes/alerter.go:1]: **Should** we ignore this in the build? It seems like a good idea to at least make sure this builds in CI.

Here and below.

File: cmd/getroughtime/main.go
Comment (cjpatton) [reply cmd/getroughtime/main.go:43]: Happy to take your suggestion.

File: cmd/getroughtime/main.go
Comment (MDr164) [reply cmd/getroughtime/main.go:74]: My spelling linter just complained about starting a log output with a lower case letter, I didn't want to change the text but I can as well do that :+1:

File: recipes/alerter.go
Comment (MDr164) [reply recipes/alerter.go:1]: Those two file only had the old build tags but as you recently bumped the Go version to 1.21 I had to add the newer build tags as gopls otherwise complained for these two files. It was previously ignored to be built but I don't know the rationale behind it.

Comment (cjpatton) [reply recipes/alerter.go:1]: No rationale, TBH we just haven't had a good maintenace story for this repo. (This code was originally written before the GitHub Actions product was generally available!)

**Comment by MDr164** [comment]
> @MDr164 this looks ready for review.

There is one more thing I wanted to add which is the integration test using a test server instance and to query it using the example client.

**Review by cjpatton — APPROVED** [review]
Looks good to me. Before merging I want to get some feedback from my teammates on the CI plan.

**Comment by MDr164** [comment]
> Looks good to me. Before merging I want to get some feedback from my teammates on the CI plan.

Sounds good. The CIRCL repo also got codecov integration, that's why I left that line a todo in the workflow yaml so you can add it if needed.

File: cmd/getroughtime/main.go
Comment (armfazh) [reply cmd/getroughtime/main.go:43]: we take most of the by-default golangci-lint linter configuration, unless it gets really picky. This is a good start, and configuration can change in the future.

**Review by armfazh — APPROVED** [review]
Changes look good, and the CI script works ok.

File: .github/workflows/build_and_test.yml
Comment (armfazh) [reply .github/workflows/build_and_test.yml:1]: script is ok.

**Pull request merged**
